text/0255-sensitive-data-redaction.md

```diff
@@ -0,0 +1,397 @@
+to add their own local requirements, either through configuration or through code.
+
+The configuration is a list of rules that the SDK can parse and apply. An item
+in this list will consist of:
+
+- Conditions when a redaction should be applied
+- Instructions for the redaction that will be applied when the conditions are met
+
+In an example configuration file a list could look like the following:
+
+```yaml
+  sensitiveData:
+      - attribute: url.query
+        redaction: REDACT_ALL_URL_VALUES
+      - attribute: com.example.customer.name
+        redaction: DROP
+      - attribute: com.example.customer.email
+        redaction: 's/^[^@]*@/REDACTED@/'
+```
+
+In this example the condition is expressed via `attribute` and the instructions
+via `redaction`, where either a constant is selected that runs a predefined helper
+method, or a sed-like expression that uses a regular expression to apply redaction.
+
+Another possibility is to re-use a language like
+[OTTL](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/README.md).
+
+### Semantic Conventions
+
+To enable the redaction requirement profiles the semantic conventions need to be
+annotated with sensitivity details.
+
+Example:
+
+| Attribute | ... existing columns ... | sensitivity details |
+|-----------|--------------------------|---------------------|
+| `url.query`|                         | Rationale: Some verbatim wording why this is the way it is below<br>Type: `mixed`<br>`DEFAULT`: `REDACT_INSECURE_URL_PARAM_VALUES`<br>STRICT: `REDACT_ALL_URL_VALUES`<br>`STRICTER`: `DROP` |
+| `client.address`|                    | Rationale: some reasons why dropping octets may be required<br>Type: `maybe_pii`<br>`DEFAULT`: `NONE`<br>`STRICT`: `'s/([0-9]+\.[0-9]+\.[0-9]+\.)[0-9]+/\10/'`<br>`STRICTER`: `'s/([0-9]+\.[0-9]+\.)[0-9]+\.[0-9]+/\10.0/'` |
+| `enduser.creditCardNumber`**[1]** |     | Rationale: ...<br>Type: `always_pii`<br>DEFAULT: `EXTRACT_IIN`<br>`STRICT`: `DROP`|
+
+**[1]**: _This is a made-up example for demonstration purpose, it’s not part of the current semantic conventions. It gives a more nuanced example, e.g. that extracting the Issuer Identification Number (IIN) might be an option over dropping the number completely. It also demonstrated the value of “type”, which can enable Data lineage use cases_
+
+The `DROP` keyword means that the value is replaced with `REDACTED` (or set to null, …).
+
+It is the responsibility of the OpenTelemetry SDK to implement those sensitivity details provided by the semantic conventions.
+
+### Annotate attributes with sensitivity information
+
+As an additional building block the OpenTelemetry API should provide capabilities
+to library (and application) authors to apply in-place redaction
+
+One option is to add additional attributes to methods like `setAttribute`:
+
+```
+span.setAttribute("url.query", url.toString(), <SENSITIVITY_DETAILS>);
```

Comment:
Would it make more sense to include sensitivity details in the *attribute itself*?

E.g.
```java
// When defining a key, we define the sensitivity under which it would be redacted.
final AttributeKey<String> myKey = AttributeKey.stringKeyWithSensitivity("url.query", Sensitivity.NO_USER, redactionFunction);
```